node/core/av-store/src/lib.rs

```diff
@@ -0,0 +1,555 @@
+					Ok(FromOverseer::Signal(_)) => (),
+					Ok(FromOverseer::Communication { msg }) => {
+						let _ = process_message(&subsystem.inner, msg);
+					}
+					Err(_) => break,
+				}
+			}
+			complete => break,
+		}
+	}
+
+	Ok(())
+}
+
+fn process_message(db: &Arc<dyn KeyValueDB>, msg: AvailabilityStoreMessage) -> Result<(), Error> {
+	use AvailabilityStoreMessage::*;
+	match msg {
+		QueryAvailableData(hash, tx) => {
+			let _ = tx.send(available_data(db, &hash).map(|d| d.data));
+		}
+		QueryDataAvailability(hash, tx) => {
+			let result = match available_data(db, &hash) {
+				Some(_) => true,
+				None => false,
+			};
+
+			let _ = tx.send(result);
+		}
+		QueryChunk(hash, id, tx) => {
+			let _ = tx.send(get_chunk(db, &hash, id)?);
+		}
+		StoreChunk(hash, id, chunk, tx) => {
+			match store_chunk(db, &hash, id, chunk) {
+				Err(e) => {
+					tx.send(Err(())).map_err(|_| oneshot::Canceled)?;
+					return Err(e);
+				}
+				Ok(()) => {
+					let _ = tx.send(Ok(()));
+				}
+			}
+		}
+		StoreAvailableData(hash, id, n_validators, av_data, tx) => {
+			match store_available_data(db, &hash, id, n_validators, av_data) {
+				Err(e) => {
+					tx.send(Err(())).map_err(|_| oneshot::Canceled)?;
+					return Err(e);
+				}
+				Ok(()) => {
+					let _ = tx.send(Ok(()));
+				}
+			}
+		}
+	}
+
+	Ok(())
+}
+
+fn available_data(db: &Arc<dyn KeyValueDB>, candidate_hash: &Hash) -> Option<StoredAvailableData> {
+	query_inner(db, columns::DATA, &available_data_key(candidate_hash))
+}
+
+fn store_available_data(
+	db: &Arc<dyn KeyValueDB>,
+	candidate_hash: &Hash,
+	id: Option<ValidatorIndex>,
+	n_validators: u32,
+	available_data: AvailableData,
+) -> Result<(), Error> {
+	let mut tx = DBTransaction::new();
+
+	if let Some(index) = id {
+		let chunks = get_chunks(&available_data, n_validators as usize)?;
+		store_chunk(db, candidate_hash, n_validators, chunks[index as usize].clone())?;
+	}
+
+	let stored_data = StoredAvailableData {
+		data: available_data,
+		n_validators,
+	};
+
+	tx.put_vec(
+		columns::DATA,
+		available_data_key(&candidate_hash).as_slice(),
+		stored_data.encode(),
+	);
+
+	db.write(tx)?;
+
+	Ok(())
+}
+
+fn store_chunk(db: &Arc<dyn KeyValueDB>, candidate_hash: &Hash, _n_validators: u32, chunk: ErasureChunk)
+	-> Result<(), Error>
+{
+	let mut tx = DBTransaction::new();
+
+	let dbkey = erasure_chunk_key(candidate_hash, chunk.index);
+
+	tx.put_vec(columns::DATA, &dbkey, chunk.encode());
+	db.write(tx)?;
+
+	Ok(())
+}
+
+fn get_chunk(db: &Arc<dyn KeyValueDB>, candidate_hash: &Hash, index: u32)
+	-> Result<Option<ErasureChunk>, Error>
+{
+	if let Some(chunk) = query_inner(
+		db,
+		columns::DATA,
+		&erasure_chunk_key(candidate_hash, index)) {
+		return Ok(Some(chunk));
+	}
+
+	if let Some(data) = available_data(db, candidate_hash) {
+		let mut chunks = get_chunks(&data.data, data.n_validators as usize)?;
+		let chunk = chunks.get(index as usize).cloned();
+		for chunk in chunks.drain(..) {
```

Comment:
This is a bit cryptic, it uses `IntoIterator` of `Option` and the for loop really only does one iteration. 

I `.map()` or `.and_then()` might be more idiomatic, or did you want to avoid the closure?

Reply:
yeah, changed the naming of the var; you see here the `Option` is `chunk` var and we are draining `chunks` which are `Vec`.